pkg/k8s/service_cache.go

```diff
@@ -499,7 +499,7 @@ func (s *ServiceCache) correlateEndpoints(id ServiceID) (*Endpoints, bool) {
 
 		for ip, e := range localEndpoints.Backends {
 			e.Preferred = svcFound && svc.IncludeExternal && svc.ServiceAffinity == serviceAffinityLocal
-			endpoints.Backends[ip] = e
```

Comment:
does this not still have the problem - even if it avoids the race - of mutating an object that we don't expect to mutate? Shouldn't you then do

```
endpoints.Backends[ip].Preferred = svcFound && svc.IncludeExternal && svc.ServiceAffinity == serviceAffinityLocal
```

Reply:
As I said above, I flipped this around and keeping the mutation of the internal state, but making sure we only hand out copies. I felt like this was the safer approach and IMO makes more sense.